# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [2]:
# Your code here
import pandas as pd
import numpy as np

url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"

df = pd.read_csv(url)

In [3]:
df.head()

Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car            2.704934  
1  Four-Door Car         1131.464935  
2   Two-Door Car          566.472247  
3            SUV          529.881344  
4  Four-Door Car           17.269323

In [4]:
df.shape

(4008, 11)

In [5]:
df.tail()

Customer   ST GENDER Education Customer Lifetime Value  Income  \
4003      NaN  NaN    NaN       NaN                     NaN     NaN   
4004      NaN  NaN    NaN       NaN                     NaN     NaN   
4005      NaN  NaN    NaN       NaN                     NaN     NaN   
4006      NaN  NaN    NaN       NaN                     NaN     NaN   
4007      NaN  NaN    NaN       NaN                     NaN     NaN   

      Monthly Premium Auto Number of Open Complaints Policy Type  \
4003                   NaN                       NaN         NaN   
4004                   NaN                       NaN         NaN   
4005                   NaN                       NaN         NaN   
4006                   NaN                       NaN         NaN   
4007                   NaN                       NaN         NaN   

     Vehicle Class  Total Claim Amount  
4003           NaN                 NaN  
4004           NaN                 NaN  
4005           NaN                 NaN  
4006           NaN                 NaN  
4007           NaN                 NaN

Jav: df has 4008 lines and 11 columns

In [6]:
df.dtypes

Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object

Jav: the types of the data are ok (float) for the numerical, but could be improved for the premium (int) 
The types of the categorical data are only set to object, they could be changed to string.
we should change all to lowercaps and replace spaces by '_'
And the units and type should be changed for CLV (from object/string to float and a realistic percent) 
Strange format for Number of Open Complaints: seems a date format

In [7]:
df.nunique()

Customer                     1071
ST                              8
GENDER                          5
Education                       6
Customer Lifetime Value      1027
Income                        774
Monthly Premium Auto          132
Number of Open Complaints       6
Policy Type                     3
Vehicle Class                   6
Total Claim Amount            761
dtype: int64

Jav: unique values analysis
customer: categorical, they repeat each other as not all are different, some have more than one insurance ?
ST: categorical, 8 states when we can see there should only be 5 => TO CORRECT 
GENDER: categorical, 5 unique values... there should be only 2, eventually 3 for the Nan    => TO CORRECT 
Education : categorical, there should be 5 and not 6 (there are 2 bachelor(s)) => TO CORRECT 
Customer Lifetime Value: categrorical wrong unit and should be numerical => TO CONVERT
Income : numerical, no comments so far (waiting on describe)
Monthly Premium Auto: numerical, 132 possible rates?, numerical, no comments so far (waiting on describe)
Number of Open Complaints:, categorical, wrong type, should be numerical => TO CORRECT
Policy Type: categorical, seems ok
Vehicle Class: categorical, seems ok
Total Claim Amount: numerical, no comments so far (waiting on describe)


In [8]:
numerical = df.select_dtypes(include='number')
categorical = df.select_dtypes(include='object')

for cat in categorical:
    print(df[cat].unique())
    print(df[cat].value_counts())

['RB50392' 'QZ44356' 'AI49188' ... 'CW49887' 'MY31220' nan]
Customer
RB50392    1
HJ15383    1
AO74776    1
HQ82233    1
OL72737    1
          ..
RO26085    1
ES57969    1
JK55587    1
RN97635    1
MY31220    1
Name: count, Length: 1071, dtype: int64
['Washington' 'Arizona' 'Nevada' 'California' 'Oregon' 'Cali' 'AZ' 'WA'
 nan]
ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64
[nan 'F' 'M' 'Femal' 'Male' 'female']
GENDER
F         457
M         413
Male       39
female     28
Femal      17
Name: count, dtype: int64
['Master' 'Bachelor' 'High School or Below' 'College' 'Bachelors' 'Doctor'
 nan]
Education
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Name: count, dtype: int64
[nan '697953.59%' '1288743.17%' ... '2031499.76%' '323912.47%'
 '899704

In [9]:
df.describe()

Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678

In [10]:
df.describe(include="object")

Customer      ST GENDER Education Customer Lifetime Value  \
count      1071    1071    954      1071                    1068   
unique     1071       8      5         6                    1027   
top     RB50392  Oregon      F  Bachelor              445811.34%   
freq          1     320    457       324                       4   

       Number of Open Complaints    Policy Type  Vehicle Class  
count                       1071           1071           1071  
unique                         6              3              6  
top                       1/0/00  Personal Auto  Four-Door Car  
freq                         830            780            576

Jav: further analysis +/- 2700 of 'empty lines
customer: categorical, they repeat each other as not all are different, some have more than one insurance ? there are 1071 unique customers and +/- 2700 of 'empty lines
ST: categorical, 8 states when we can see there should only be 5 => TO CORRECT 
    Seems there are about a third of people from Oregon, but the data has not yet been cleaned 
GENDER: categorical, 5 unique values... there should be only 2, eventually 3 for the Nan    => TO CORRECT 
    More female? to correct the data first
Education : categorical, there should be 5 and not 6 (there are 2 bachelor(s)) => TO CORRECT 
    More bachelor(s), even without correcting yet! (+/- 30%)
Customer Lifetime Value: categrorical wrong unit and should be numerical => TO CONVERT
    very different values, to change to numerical for real analysis..
Income : numerical, no comments so far (waiting on describe)
    Outliers: income of 0... is very low, 100K$ for the max is not out of range. 50% of the insured are within 14-65K$
    Average and median are close, it seems a normal distribution, slightly right skewed (more 'insured' to the left)
Monthly Premium Auto: numerical, 132 possible rates?, numerical, no comments so far (waiting on describe)
    Outliers: very strange maximum, extremely high, a lot of low values between 61-68 USD
    average and median are far appart: distribution stronly right skewed. 
    The high values have really to be checked, the std value is high, there are lots of differences between the 'normal' and 'high values'
Number of Open Complaints:, categorical, wrong type, should be numerical => TO CORRECT
    to convert before further analysis
Policy Type: categorical, seems ok
    Mainly Personal Auto is used, +/- 75% of the time
Vehicle Class: categorical, seems ok
    Mainly Four-Door Car is used, +/-55% of the time
Total Claim Amount: numerical, every body has claimed something (minimum > 0)
    Outliers: not obvious
    Average and median: +/- normal distribution
    50% between 200-530

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [11]:
# Your code here
ser_locations = df['ST']
ser_locations.value_counts(ascending = True)

ST
AZ             25
WA             30
Washington     81
Nevada         98
Cali          120
Arizona       186
California    211
Oregon        320
Name: count, dtype: int64

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [12]:
# Your code here

ser_policies = df['Policy Type']
ser_policies.value_counts()

Policy Type
Personal Auto     780
Corporate Auto    234
Special Auto       57
Name: count, dtype: int64

In [13]:
ser_policies.loc[max(ser_policies.value_counts())]

'Personal Auto'

### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [38]:
# Your code here

# filter for 'Personal Auto', and describe gives the stats for the numerical columns
mean_perso_auto = df[df['Policy Type'].isin(['Personal Auto'])]["Income"].mean() 
mean_corpo_auto = df[df['Policy Type'].isin(['Corporate Auto'])]["Income"].mean() 

print(f' Personal Auto Income Mean = {mean_perso_auto}')
print(f'Corporate Auto Income Mean = {mean_corpo_auto}')



 Personal Auto Income Mean = 38180.69871794872
Corporate Auto Income Mean = 41390.31196581197


Jav: yes Corporate Auto Income Mean >  Personal Auto Income Mean

### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [45]:
# Your code here
df["Total Claim Amount"].value_counts()


Total Claim Amount
321.600000    19
331.200000    15
312.000000    14
350.400000    14
316.800000    12
              ..
421.484456     1
430.505942     1
7.646763       1
55.510526      1
207.320041     1
Name: count, Length: 761, dtype: int64

In [47]:
# My way...
sorted_df = df.sort_values("Total Claim Amount")

sorted_df.iloc[:190].describe() # 761/4 = 190.25



Income  Monthly Premium Auto  Total Claim Amount
count    190.000000            190.000000          190.000000
mean   54347.089474            340.247368           63.646148
std    25651.827099           2636.571743           36.317637
min        0.000000             61.000000            0.382107
25%    36109.500000             65.250000           32.114902
50%    53951.500000             73.000000           61.731668
75%    73843.750000             99.000000           92.889787
max    99428.000000          35354.000000          132.588288

In [49]:
#another way is using threshold = df['col_name'].quantile(0.75) then a conditional filter
# the data is not exactly the same as it takes the values from the quantile, so there are more than in my own calculation...

threshold = df['Total Claim Amount'].quantile(0.75)
filtered_df = df[df['Total Claim Amount'] > threshold]

filtered_df.describe()

Income  Monthly Premium Auto  Total Claim Amount
count    264.000000            264.000000          264.000000
mean   23677.344697            165.193182          782.228263
std    27013.483721            623.930992          292.751640
min        0.000000             63.000000          537.600000
25%        0.000000             99.000000          606.521741
50%    18807.000000            114.000000          679.597985
75%    42423.750000            133.250000          851.400000
max    99316.000000          10202.000000         2893.239678

In [50]:
filtered_df.describe(include='object')

Customer      ST GENDER Education Customer Lifetime Value  \
count       264     264    238       264                     264   
unique      264       7      5         5                     256   
top     QZ44356  Oregon      F  Bachelor              578018.22%   
freq          1      90    115        85                       3   

       Number of Open Complaints    Policy Type Vehicle Class  
count                        264            264           264  
unique                         6              3             6  
top                       1/0/00  Personal Auto           SUV  
freq                         206            191           101

Jav: for the top 25% of the Total Claim Amount the data is as follows
Customer 264 of them
ST : 7 states? (in fact should be 5)
GENDER 5, still weird, should be only 2
Education 5 levels of education, Bachelors in majority a bit below 30%
Customer Lifetime Value: difficult to interpret
Income : numerical, about 10K less than the total population (looking at the mean), also skewed right
Monthly Premium Auto         101
Number of Open Complaints      6
Policy Type, 3, mainly Personal Auto, for a bit more than 70% (191/264)
Vehicle Class, all the 6, close to 40% of SUV!!
Total Claim Amount : as expected much more than the total population as it is the top percentile.

In [52]:
df.describe()

Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678